File: go-backend-interview-prep/08_http_and_backend/10_graceful_shutdown/main.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net"
	"net/http"
	"os"
	"os/signal"
	"syscall"
	"time"
)

func writeJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(v)
}

// slowHandler simulates a request that takes time to complete.
func slowHandler(w http.ResponseWriter, r *http.Request) {
	fmt.Println("  [handler] slow request started, sleeping 2s...")
	time.Sleep(2 * time.Second)
	writeJSON(w, http.StatusOK, map[string]string{
		"message": "slow request completed",
	})
	fmt.Println("  [handler] slow request done")
}

func healthHandler(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, http.StatusOK, map[string]string{"status": "ok"})
}

func main() {
	mux := http.NewServeMux()
	mux.HandleFunc("/health", healthHandler)
	mux.HandleFunc("/slow", slowHandler)

	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		log.Fatal(err)
	}
	addr := ln.Addr().String()

	server := &http.Server{Handler: mux}

	// --- Graceful shutdown on signal ---
	quit := make(chan os.Signal, 1)
	signal.Notify(quit, syscall.SIGINT, syscall.SIGTERM)

	go func() {
		sig := <-quit
		fmt.Printf("\n  [shutdown] received signal: %s\n", sig)
		fmt.Println("  [shutdown] draining in-flight requests (10s timeout)...")

		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
		defer cancel()

		if err := server.Shutdown(ctx); err != nil {
			fmt.Printf("  [shutdown] error: %v\n", err)
		} else {
			fmt.Println("  [shutdown] graceful shutdown complete")
		}
	}()

	fmt.Printf("Server on http://%s\n", addr)
	fmt.Println("  GET /health -> health check")
	fmt.Println("  GET /slow   -> 2s slow request")
	fmt.Println("  Press Ctrl+C to trigger graceful shutdown")

	// Self-test: start a slow request, then trigger shutdown mid-flight
	go func() {
		time.Sleep(200 * time.Millisecond)
		base := "http://" + addr
		fmt.Println("\n--- Demo ---")

		// Start slow request in background
		done := make(chan struct{})
		go func() {
			defer close(done)
			resp, err := http.Get(base + "/slow")
			if err != nil {
				fmt.Printf("  slow request error: %v\n", err)
				return
			}
			var body map[string]string
			json.NewDecoder(resp.Body).Decode(&body)
			resp.Body.Close()
			fmt.Printf("  slow request result: %v (status %d)\n", body, resp.StatusCode)
		}()

		// Wait a bit then signal shutdown while slow request is in-flight
		time.Sleep(500 * time.Millisecond)
		fmt.Println("  [demo] sending SIGINT while slow request is in-flight...")
		quit <- syscall.SIGINT

		// Wait for slow request to complete (should finish under graceful drain)
		<-done
	}()

	// Serve blocks until Shutdown is called
	if err := server.Serve(ln); err != http.ErrServerClosed {
		log.Fatal(err)
	}

	// Wait a moment for shutdown goroutine to finish logging
	time.Sleep(100 * time.Millisecond)
	fmt.Println("\nServer exited.")
}

File: go-backend-interview-prep/08_http_and_backend/10_graceful_shutdown/main.py
"""Graceful shutdown -- Python equivalent of the Go example.

Python's HTTPServer.shutdown() stops serve_forever() but doesn't drain
in-flight requests as gracefully as Go's Shutdown(ctx). We demonstrate
signal handling and the best stdlib approach.
"""

import json
import signal
import threading
import time
import urllib.request
from http.server import HTTPServer, BaseHTTPRequestHandler


def json_response(handler, status: int, data: dict) -> None:
    handler.send_response(status)
    handler.send_header("Content-Type", "application/json")
    handler.end_headers()
    handler.wfile.write(json.dumps(data).encode())


class Handler(BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path == "/health":
            json_response(self, 200, {"status": "ok"})
        elif self.path == "/slow":
            print("  [handler] slow request started, sleeping 2s...")
            time.sleep(2)
            json_response(self, 200, {"message": "slow request completed"})
            print("  [handler] slow request done")
        else:
            json_response(self, 404, {"error": "not found"})

    def log_message(self, *args):
        pass


def main() -> None:
    server = HTTPServer(("127.0.0.1", 0), Handler)
    host, port = server.server_address
    base = f"http://{host}:{port}"

    # --- Signal handler for graceful shutdown ---
    def on_signal(signum, frame):
        name = signal.Signals(signum).name
        print(f"\n  [shutdown] received signal: {name}")
        print("  [shutdown] stopping server...")
        # shutdown() must run in a separate thread (it blocks until serve_forever stops)
        threading.Thread(target=server.shutdown).start()

    signal.signal(signal.SIGINT, on_signal)
    signal.signal(signal.SIGTERM, on_signal)

    print(f"Server on {base}")
    print("  GET /health -> health check")
    print("  GET /slow   -> 2s slow request")
    print("  Press Ctrl+C to trigger graceful shutdown")

    # Self-test demo
    def demo():
        time.sleep(0.2)
        print("\n--- Demo ---")

        # Start slow request in background
        result = [None]

        def slow_req():
            try:
                resp = urllib.request.urlopen(f"{base}/slow", timeout=10)
                result[0] = (resp.status, json.loads(resp.read()))
            except Exception as e:
                result[0] = ("error", str(e))

        t = threading.Thread(target=slow_req)
        t.start()

        # While slow request is in-flight, trigger shutdown
        time.sleep(0.5)
        print("  [demo] triggering shutdown while slow request is in-flight...")
        threading.Thread(target=server.shutdown).start()

        # Wait for slow request
        t.join()
        if result[0]:
            print(f"  slow request result: {result[0]}")

    threading.Thread(target=demo, daemon=True).start()

    # serve_forever blocks until shutdown() is called
    server.serve_forever()
    print("  [shutdown] server stopped")
    print("\nServer exited.")


if __name__ == "__main__":
    main()
